From d2901a2e08513543d11789774fc8a548291b09ad Mon Sep 17 00:00:00 2001
From: Nadav Haklai <nadavh@marvell.com>
Date: Tue, 8 Sep 2015 13:28:58 +0300
Subject: [PATCH 0021/1345] ARM: mvebu: Enable SCU Speculative linefills to L2

commit  f77997b21f5e6f658d9b99ae9d25551a35db214a from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

When set, coherent linefill requests are sent speculatively to the L2C-310
in parallel with the tag look-up. If the tag look-up misses, the confirmed
linefill is sent to the L2C-310 and gets RDATA earlier because the data
request was already initiated by the speculative request.

This feature may improve overall system performance.

Change-Id: Ib7b29668bace0f09762a7cbab3ea149bc776891c
Signed-off-by: Nadav Haklai <nadavh@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/23491
Reviewed-by: Neta Zur Hershkovits <neta@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/23672
Reviewed-on: http://vgitil04.il.marvell.com:8080/27307
Tested-by: Lior Amsalem <alior@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-mvebu/board-v7.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/arch/arm/mach-mvebu/board-v7.c b/arch/arm/mach-mvebu/board-v7.c
index 1aa3fd6..3c58a7f 100644
--- a/arch/arm/mach-mvebu/board-v7.c
+++ b/arch/arm/mach-mvebu/board-v7.c
@@ -33,6 +33,7 @@
 #include "coherency.h"
 #include "mvebu-soc-id.h"
 
+#define SCU_CTRL		0x00
 static void __iomem *scu_base;
 
 /*
@@ -41,10 +42,19 @@
  */
 static void __init mvebu_scu_enable(void)
 {
+	u32 scu_ctrl;
 	struct device_node *np =
 		of_find_compatible_node(NULL, NULL, "arm,cortex-a9-scu");
 	if (np) {
 		scu_base = of_iomap(np, 0);
+
+		scu_ctrl = readl_relaxed(scu_base + SCU_CTRL);
+		/* already enabled? */
+		if (!(scu_ctrl & 1)) {
+			/* Enable SCU Speculative linefills to L2 */
+			scu_ctrl |= (1 << 3);
+			writel_relaxed(scu_ctrl, scu_base + SCU_CTRL);
+		}
 		scu_enable(scu_base);
 		of_node_put(np);
 	}
-- 
1.7.9.5

